### DISCUSSION TITLE
Migrate JavaScript development to Eglot from Emacs LSP : Node, Deno, and Monorepos

### DISCUSSION THREAD

**Discussion by cowboyd** [joaotavora/eglot#1337] [General]
Hello,

My Holiday project is to take a stab at migrating from emacs-lsp to eglot for JavaScript development. Unfortunately, I've had a pretty rough go of it so far. The problem I'm having is that the repos I'm operating in have multiple JavaScript projects, some of which use Node as the primary development tool, others which use Deno as the primary development tool. 

I need (I think) for Eglot to recognize each individual javascript package as a separate project apart from the main vc-project. Furthermore, if the package has a `deno.json` file, then use the `deno lsp` command, otherwise, use the `typescript-language-server` command.

Thus far, I've tried to follow what I can piece together from the documentation and [this discussion on using an alternate project root](https://github.com/joaotavora/eglot/discussions/687) to add a hook to `project-find-functions`. Here's my set up so far

```lisp
(defun @cowboyd/find-nearest-dominating-file (files dir)
  "Find first one of `files' searching upwards from `dir' "  
  (if (string= dir "/") nil
    (let* ((contents (directory-files dir))
	  (file (seq-find (lambda (file) (seq-contains-p contents file)) files)))
      (or
       (when file (concat (file-name-as-directory dir) file))
       (@cowboyd/find-nearest-dominating-file files (directory-file-name (file-name-directory dir)))))))

;; get nearest Deno project file that does not have a Node project file in front of it.
(defun @cowboyd/deno-project (&optional startdir)
  "Is the current file inside a deno (and not a node) project. If a 'package.json' is found first, then this is not a deno directory"
  (let* ((dir (directory-file-name (or startdir (file-name-directory (buffer-file-name)))))
	(project-file (@cowboyd/find-nearest-dominating-file
		       (list "deno.json" "package.json")
		       dir)))
    (when (string= "deno" (file-name-base project-file)) project-file)))

(defun @cowboyd/deno-eglot-project (dir)
  "Return a deno directory scoped project root for eglot if this file is in a deno directory"
  (and (boundp 'eglot-lsp-context) eglot-lsp-context ;;; short circuit if this is not eglot searching for project
    (let ((deno-config (@cowboyd/deno-project dir)))
      (when deno-config (cons 'transient (file-name-directory deno-config))))))
  
(add-hook 'project-find-functions '@cowboyd/deno-eglot-project)
```
However, as far as I can tell whenever I switch to the main project and find a file located in the sub-package, `(project-current)` is already computed for the buffer and assigned somewhere, and so my custom find function is never run. As a result, Eglot retrieves the root vc directory instead of the specific directory.

It does appear to get the right directory if I open the sub-package file directly.

Finally, given that I have a way to detect between Deno projects and Node projects, what is the best way to get Eglot to select the right language server dynamically?

emacs-lsp solves this problem by letting you explicitly state via a config file that a directory is a workspace root. I haven't seen this as an option in Eglot (I could be wrong). But anyway,  I've been looking to migrate over to eglot for some time, and so any insight would be very much appreciated.

**Comment by joaotavora** [comment]
I'll try to read this later, as one of my holidays projects is resting, but seems very similar to https://github.com/joaotavora/eglot/discussions/1336

**Reply by cowboyd** [reply]
Thanks! I did read that discussion as well as the one that I linked. There are two problems I'm facing: the non-vc-root and then dynamically choosing Deno vs Node. Do you know of how to solve the latter?

Enjoy your holiday and feel free to not respond anymore 🍻!!

**Reply by joaotavora** [reply]
So the two problems here:

> 1.  the non-vc-root

This is solved by the strategy you were attempting, but only if a long-standing bug didn't exist in this feature which makes it really a pitfall.  Technical detail `eglot-lsp-context`  isn't bound, it's just declared special via a valueless top-level `defvar`. Not sure what I was thinking: it should just be nil globally.  

Until i fix this in Eglot (which I've just done, so you may try to get the bleeding edge  from GNU-devel ELPA)  a simple `(setq eglot-lsp-context nil)` will likely fix your snippets in that respect.  The snippet will no longer have to use `boundp` and can just query the variable's value. 

(Unrelated: by the way recommend against `seq-find` (and `seq-` functions in general). `locate-dominating-file` should do what you want here)

Note that the project returned by Eglot is of the `transient` kind which AFAIK is the only type of project libraries are allowed to make easily.  It will work, but will be much much slower for getting the project files, for example.  Eglot doesn't do that very often though (I think only once at startup, for executing the `workspace/didChangeWatchedFiles` request), so you may be safe.

> 2. The deno vs typescript-langauge-server thing

I would just use a directory-local value for `eglot-server-programs`.  See in the Emacs manual how to do that.  You can do it with files in the project, or with Elisp in your config.

Or more likely I would not even bother with this and just `C-u M-x eglot`, look up in the history (usually via the omnipresent `C-r`), for `deno` and run that.  

I suppose you want to use `eglot-ensure` right to have everything done automatically?  Well I _also_ recommend against that.  Just in case you missed it in 

> The function eglot-ensure will start an Eglot session for each buffer in which foo-mode is turned on, if there isn’t already an Eglot session that handles the buffer. Note that this variant of starting an Eglot session is non-interactive, so it should be used only when you are confident that Eglot can be started reliably for any file which may be visited with the major-mode in question.

> Note that it’s often difficult to establish this confidence fully, so it may be wise to use the interactive command eglot instead.  --->**You only need to invoke it once per project** <---, as all other files visited within the same project will automatically be managed with no further user intervention needed.

(emphasis mine for a detail usually missed by users migrating over from lsp-mode)

**Reply by joaotavora** [reply]
Anyway, I don't know what you mean by "in front", but this untested hack certainly _looks_ simpler than your solution, right?
```lisp
(defmacro my/define-eglot-subproject-marker (name marker)
  `(defun ,name (dir)
     (when-let ((match
                 (and eglot-lsp-context
                      (locate-dominating-file dir ,marker))))
       (cons 'transient match))))

(my/define-eglot-subproject-marker my/find-deno "deno.json")
(my/define-eglot-subproject-marker my/find-pjson "package.json")

;; pjson before deno, and both before .git
(add-hook 'project-find-functions #'my/find-pjson -10)
(add-hook 'project-find-functions #'my/find-deno -5)
```

**Reply by cowboyd** [reply]
Thanks for the tips. I'm very close now I think. I'm trying to follow your advice and just starting eglot interactively until I can feel confident in an automated solution (if that moment ever arrives).

I'm trying to set up to be able to select either typescript or Deno using `eglot-alternatives`. Following [this setup from the Deno homepage](https://docs.deno.com/runtime/manual/getting_started/setup_your_environment#eglot)  which says to create a custom eglot subclass for deno like so:.

```elisp
  (defclass eglot-deno (eglot-lsp-server) ()
    :documentation "A custom class for deno lsp.")

  (cl-defmethod eglot-initialization-options ((server eglot-deno))
    "Passes through required deno initialization options"
    (list :enable t
    :lint t))
```
And this works when it is the only eglot server program, but in order to add `typescript-language-server` back so that I can choose between the two, I add it to my `eglot-server-programs` using `eglot-alternatives`

```elisp
(add-to-list 'eglot-server-programs
	     '((js-mode (typescript-ts-base-mode :language-id "typescript")) .
	       ,(eglot-alternatives
		'(
		  ("typescript-language-server" "--stdio")
		  (eglot-deno "deno" "lsp")))))
```

But I don't think I'm doing it right because when I try and run eglot in a typescript buffer, I get this error:

```txt
cl-no-applicable-method: No applicable method: make-instance, eglot-alternatives, :name, "EGLOT (effection/(typescript-ts-mode js-mode typescript-ts-base-mode))", :events-buffer-scrollback-size, 2000000, :notification-dispatcher, #[771 "����\"$)�" [eglot-handle-notification eglot--cached-server apply append nil] 10 "

(fn SERVER METHOD PARAMS)"], :request-dispatcher, #[771 "����\"$)�" [eglot-handle-request eglot--cached-server apply append nil] 10 "

(fn SERVER METHOD PARAMS)"], :on-shutdown, eglot--on-shutdown
```

As for the non-vc-root. The problem is that you can have a directory layout like this:

```
- parent/
| - package.json
+ - subdir/
  | - deno.json
  | - file.ts
```

And so if `project-find-functions` looks for `package.json` first and gets a hit, it would incorrectly choose `parent/` as the project directory.

**Reply by joaotavora** [reply]
Here's one way I think it can be done.  The caching might be overkill.  In theory, if project.el supported subprojects, you wouldn't need it.  And maybe you don't need it if you don't mind traversing up-tree twice for each file.

```lisp
(defun my/find-marker ()
  (or my/marker
      (setq my/marker
            (catch 'done
              (locate-dominating-file
               default-directory
               (lambda (d)
                 (dolist (f '("package.json" "deno.json"))
                   (let ((marker (expand-file-name f d)))
                     (when (file-exists-p marker) (throw 'done marker))))))))))

(defun my/deno-or-tss ()
  (lambda (&optional _interactive)
    (let* ((m (my/find-marker)))
      (cdr
       (assoc
        (file-name-nondirectory m)
        '(("package.json" "typescript-language-server" "--stdio")
          ("deno.json" "deno" "lsp" :initializationOptions '(:enable t :lint t))))))))

(defun my/project-find-function (dir)
  (when-let ((f (and eglot-lsp-context (my/find-marker))))
    `(transient . ,(file-name-directory f))))

(add-hook 'project-find-functions 'my/project-find-function)
(add-to-list 'eglot-server-programs
             `((js-mode (typescript-ts-base-mode :language-id "typescript")) .
               ,(my/deno-or-tss)))
```

**Reply by dgutov** [reply]
project.el supports subprojects using `project-vc-extra-root-markers`.

I'm surprised you don't remember that.

**Reply by joaotavora** [reply]
@cowboyd  (presumably) wants these smaller projects for Eglot connection purposes.   If they use your technique, `C-x p f` and a lot of other commands becomes useless for jumping between files of different modules in the overarching project.

Even if @cowboyd doesn't want to preserve that specifically in this instance, I do.  So  Eglot-specific subprojects are better than no subproject support at all in `project.el`.  It that existed in some useful form, this probably wouldn't be needed.

> I'm surprised you don't remember that.

I'm surprised you didn't notice my explicit ping here: https://github.com/joaotavora/eglot/discussions/1336#discussioncomment-7911270, where I told that other user there are easy ways to control projects Emacs-wide:

> > You tell your Emacs session as a whoel that the root of the project is not the root of the repo. -- for that particular project. That affects everything project related, not just Eglot.
> > [...]
> > Maybe there's an easy way to do the first. Ask the project.el manager @dgutov .

**Reply by joaotavora** [reply]
I didn't tell the user about `project-vc-extra-root-markers` specifically because I couldn't remember the name, I don't exactly know how they work, and I suspected they would be rather useless (at least for me) with the above limitation.  I seem to have confirmed my suspicion.  If I put `(setq project-vc-extra-root-markers '("package.json"))` and have a large hierarchy and `project.json` appears in some intermediate dir.  I can't use `C-x p f` to jump to files outside that intermediate dir from directories inside that intermediate dir.  

EDIT: even stranger is that if I jump from the outside project to the inner one with `C-x p f`, I'm trapped, can't jump back.

**Reply by dgutov** [reply]
> I didn't tell the user about project-vc-extra-root-markers specifically because I couldn't remember the name

`M-x customize-group` is your friend. No need to call me in for simple questions every time.

> I can't use C-x p f to jump to files outside that intermediate dir from directories inside that intermediate dir.

You either can't use that, or you can't use it to jump across the narrowed "subproject" only, which is also useful. The choice between the two is arbitrary, and it's not obvious which one is more useful in general. At least the current way allows "switching" to top-level project by going up a few dirs.

> even stranger is that if I jump from the outside project to the inner one with C-x p f, I'm trapped, can't jump back

The users who I asked previously said that this is okay and in some ways preferable. Projectile does this too, for example.

Would you prefer that the "outside" project excluded subproject files? That would be "consistent", but would it be more useful?

Anyway, I'm sure you know the appropriate venue for reporting bugs and requesting features.

**Reply by joaotavora** [reply]
> M-x customize-group is your friend. No need to call me in for simple questions every time.

Why bother? Just because I didn't know the name didn't prevent me from correctly it would be useless, as has been shown. But I'll leave you alone.  Just don't be "suprised" 

> The users who I asked previously said that this is okay and in some ways preferable. Projectile does this too, for example.

Preferable to be stuck?  

> Would you prefer that the "outside" project excluded subproject files?

No.  I would prefer subprojects to be useful in navigating for large monorepos.

> Anyway, I'm sure you know the appropriate venue for reporting bugs and requesting features.

Don't hold your breath. I've already done that, and even supplied a patch.

**Comment by cowboyd** [comment]
When it came to auto selecting typescript or deno, this worked like a charm! Now I just need to figure out how to parse it in my mind. In particular, I haven't yet grokked this expression:

```elisp
`((js-mode (typescript-ts-base-mode :language-id "typescript")) .
               ,(my/deno-or-tss))
```

But I'm keen to figure it out.

While the lsp connection worked for individual node and deno projects, I was not able to get the `project-find-function` to work as it keeps giving an error in all contexts:

```
Symbol's value as variable is void: eglot-lsp-context
```

**Reply by joaotavora** [reply]
>  I haven't yet grokked this expression:

That's just creating a function instead of a list of strings, which is a legal object according to the docstring of `eglot-server-programs`.  The function of one argument is returned by a call to `(my/deno-or-tss)`.  There are other ways to do it, like putting the lambda there directly.  In fact, `eglot-alternatives` is another example of such a function-returning function.

> Symbol's value as variable is void: eglot-lsp-context

You have to have either the latest bleeding edge eglot or put `(setq eglot-lsp-context nil)` somewhere in there.

**Reply by joaotavora** [reply]
> That's just creating a function instead of a list of strings, which is a legal object according to the docstring of eglot-server-programs. The function of one argument is returned by a call to (my/deno-or-tss). There are other ways to do it, like putting the lambda there directly. In fact, eglot-alternatives is another example of such a function-returning function.

I was overcomplicating this.  If you want, erase the above from your head, and use this instead, which is easier to grok:

```lisp
(defun my/find-marker ()
  (or my/marker
      (setq my/marker
            (catch 'done
              (locate-dominating-file
               default-directory
               (lambda (d)
                 (dolist (f '("package.json" "deno.json"))
                   (let ((marker (expand-file-name f d)))
                     (when (file-exists-p marker) (throw 'done marker))))))))))

(defun my/deno-or-tss-v2 (&optional interactive)
  (let* ((m (my/find-marker)))
    (cdr
     (assoc
      (file-name-nondirectory m)
      '(("package.json" "typescript-language-server" "--stdio")
        ("deno.json" "deno" "lsp" :initializationOptions '(:enable t :lint t)))))))

(defun my/project-find-function (dir)
  (when-let ((f (and eglot-lsp-context (my/find-marker))))
    `(transient . ,(file-name-directory f))))

(add-hook 'project-find-functions 'my/project-find-function)
(add-to-list 'eglot-server-programs
             '((js-mode (typescript-ts-base-mode :language-id "typescript"))
               . my/deno-or-tss-v2))
```

I will be working to simplify this further in the bleeding-edge Eglot, and maybe put some recipe in the manual

**Reply by joaotavora** [reply]
In the very latest Eglot, you can now do this.  Take note, however that the 
project type represented by `(eglot--project ROOT KEY1 VAL1 KEY2 VAL2)` is still experimental
and so is this whole technique.  But I won't remove support for it unless we get something much better.

I'm still working out how this fits with the existing project-specific and user-specific configuraiton.

```lisp
(defun my/deno-or-tss-v3 (interactive project)
  (let ((marker (and (eq (car project) 'eglot--project)
                     (plist-get (cddr project) :marker))))
    (cond ((string= marker "package.json")
           '("typescript-language-server" "--stdio"))
          ((string= marker "deno.json")
           '("deno.json" "deno" "lsp" :initializationOptions '(:enable t :lint t)))
          (interactive
           (message "Can't guess server invocation for '%s'" (project-root project))
           nil))))

(defun my/project-find-function (dir)
  (let* (marker
         (root (locate-dominating-file
                default-directory
                (lambda (d)
                  (dolist (f '("package.json" "deno.json"))
                    (when (file-exists-p (expand-file-name f d))
                      (setq marker f)))))))
    (when root
      ;; Use experimental eglot--project project type which has some
      ;; space for user-defined props, `:marker' in this case
      `(eglot--project root :marker marker))))

(add-hook 'project-find-functions 'my/project-find-function)
(add-to-list 'eglot-server-programs
             '((js-mode (typescript-ts-base-mode :language-id "typescript"))
               . my/deno-or-tss-v3))
```

**Comment by cowboyd** [comment]
> @cowboyd (presumably) wants these smaller projects for Eglot connection purposes. If they use your technique, C-x p f and a lot of other commands becomes useless for jumping between files of different modules in the overarching project.

☝🏻 This is actually what caused me to seek help in the first place. Using `C-x p f` was always assigning the wrong project to typescript files in the deno-based sub-folder.

As far as retaining the ability to have your own concept of an eglot project, I think it's wise. One of the things that emacs-lsp gets right is that it keeps its own conception of what and where a project is and as a result getting started is really straightforward. It just makes on-boarding from wherever a user happens to be much more friction-free.

Apart from that, I must say that I am really enjoying eglot thus far. I just love how well it slides into all the other packages which I'm already using rather than maintain its own parallel ecosystem of everything. Nice work!

Question: Before I decide on whether to go "bleeding edge. " How often do you release "stable" versions of eglot to ELPA?

**Reply by joaotavora** [reply]
> rather than maintain its own parallel ecosystem of everything. Nice work!

That's the general idea yes. Glad you like it.

> As far as retaining the ability to have your own concept of an eglot project, I think it's wise.

Yes, retaining this ability is important, but if "subprojects" had better support in `project.el`, maybe Eglot wouldn't need to exercise it as often, as it could just hook onto the inner project by default, the inner project presumably being what the hypothetical `project-current` would return either by default or when given a ':inner t' argument somehow.   `C-x f p` and other tools would maybe default to the outer project, like they do now.  Then maybe, in some (most?) cases, an Eglot specific concept wouldn't be required.  

Not sure how `vc-extra-markers` fit into the picture, they don't seem very useful now (at least for my own megarepo over at $DAYJOB, where marker files are unfeasible).

Maybe all of the above or equivalent will happen, but I I wouldn't hold my breath.

> How often do you release "stable" versions of eglot to ELPA?

Not very often I'm afraid.   You can help me maintain this practice, by pinging me, testing and keeping the NEWS file up-to-date :-).  Anyway the bleeding edge is what I use myself.

**Comment by cowboyd** [comment]
@joaotavora I've finally got around to trying out the bleeding edge and subprojects (I've been limping along and mostly been fine, but enough was enough)

I had to adapt it since it wasn't quite working. I had to unquote `root` and `marker` in the project definition... this seems to work?

```elisp
(defun @cowboyd/deno-lsp-or-typescript-language-server (&optional interactive project)	 
  (let ((marker (and (eq (car project) 'eglot--project)
                   (plist-get (cddr project) :marker))))
	   (cond ((string= marker "package.json")
		  '("typescript-language-server" "--stdio"))
		 ((string= marker "deno.json")
		  '("deno" "lsp" :initializationOptions '(:enable t :lint t)))
		 (interactive
		  (message "Can't guess server invocation for '%s'" (project-root project))
		  nil))))

(defun @cowboyd/js-project-find-function (dir)
  (let* ((marker nil)
         (root (catch 'done
		 (locate-dominating-file
                  default-directory
                  (lambda (d)
                    (dolist (f '("package.json" "deno.json"))
                      (when (file-exists-p (expand-file-name f d))
			(setq marker f)
			(throw 'done (file-name-parent-directory (expand-file-name f d))))))))))

    (when root
      ;; Use experimental eglot--project project type which has some
      ;; space for user-defined props, `:marker' in this case
      `(eglot--project ,root :marker ,marker))))

(add-hook 'project-find-functions '@cowboyd/js-project-find-function)
(add-to-list 'eglot-server-programs
             `((js-mode (typescript-ts-base-mode :language-id "typescript"))
	       . @cowboyd/deno-lsp-or-typescript-language-server))
```

Also, note that I had to rejigger the `let*` to throw the root directory since it was never getting set before.

So far, so good. I'll let you know if anything changes.

**Reply by cowboyd** [reply]
Everything seems great, however whenever I am in a `(eglot--project)` I'm not able to switch out of it using `(project-switch-project)` which is pretty core to my workflow. Basically what happens is that when I select a project to switch to, it just returns be back to the eglot project

**Reply by cowboyd** [reply]
Another thing I've noticed is that there is a very significant performance degradation when finding files with `(project-find-file)` in the larger javascript monorepo that is all under a single project. Sometimes it can be several seconds between when `(project-find-file)` is invoked and when results are shown.

**Reply by dgutov** [reply]
> a very significant performance degradation

I suppose that's due to your project definition being applied generally, and not just for eglot (when it asks for it).

A general-purpose project backend needs to have an optimized `project-files` implementation; quite a bit of effort went into the one that the project-vc backend has.

**Reply by joaotavora** [reply]
> I suppose that's due to your project definition being applied generally, and not just for eglot (when it asks for it).

That's a possibility yes, seems like the `eglot-lsp-context`  isn't being taken in consideration anywhere (not in my examples, nor in @cowboyd's)

> A general-purpose project backend needs to have an optimized project-files implementation; quite a bit of effort went into the one that the project-vc backend has.

As far as I can tell, it's fast for small projects, but rather useless for larger ones.  A better performing implementation for larger repos should use external-completion.el, not attempt to transfer all the possible file names into the Emacs address space and find some specialized external tool that keeps an index (like `everything.exe` on Windows, `locate` or `fdfind` on linux, or some custom-made tool just for Emacs). Much like symbol completions, you'll lose styles but you'll gain heaps in responsiveness.  Just an idea.

**Reply by dgutov** [reply]
`project-file-completion-table` was the "canonical" generic at some point, see e.g. commit 55ec674f5090. IIRC that made certain things more difficult, and induced more copying (thus making things uniformly slower for implementations that don't use external storage).

Basically, if you want to show file completions even for empty input, this is where the problem starts. Could it be solved with a combination of a streaming interface *and* external filtering? Maybe.

**Comment by joaotavora** [comment]
> I had to adapt it since it wasn't quite working. I had to unquote root and marker in the project definition... this seems to work?

unquoting is obvisouly needed yes.  I was going to swear I tested that code, but this means I haven't.

> Everything seems great, however whenever I am in a (eglot--project) I'm not able to switch out of it using (project-switch-project) which is pretty core to my workflow. Basically what happens is that when I select a project to switch to, it just returns be back to the eglot project

This would need a repro, so we can edebug it.

> Another thing I've noticed is that there is a very significant performance degradation when finding files with (project-find-file) in the larger javascript monorepo that is all under a single project. Sometimes it can be several seconds between when (project-find-file) is invoked and when results are shown.

Does this happen without any subprojects at all?  Or only when you plug in your solution?

If the former, I can tell you I also work with a (veeery large 600 K files, larger than yours? 🤡 ) and `M-x project-find-file` is positively useless there, even with the Git stuff it leverages.  A better alternative would to use something like Everything.exe for windows (https://www.voidtools.com/downloads/) (or, in linux, maybe something equivalent like locatedb) and `external-completion.el` for responsiveness.  I'm using this quick hack to contact an Everything.exe server: https://github.com/joaotavora/eel.  It's fairly performant, and above all, responsive.

It's _not_ plugged into `project.el`'s `C-x f` though, so I just bind a key.  No idea if it can even be plugged, probably yes.  project.el seems very VC/Git-centric, but who knows maybe you can just feed it a completion table.

**Reply by dgutov** [reply]
> I also work with a (veeery large 600 K files, larger than yours? 🤡 ) and M-x project-find-file is positively useless there

There is a certain threshold for the project size where the delay becomes too much, that depends on the person and the machine, though. With a 200K project I'm testing, `project-find-file` takes about 1 second, which is not great but might be considered manageable by some. There is still some optimization space left, but after that, I guess it's back to caching the file list, like Projectile does. Or external databases like in your approach.

**Reply by joaotavora** [reply]
> I guess it's back to caching the file list

You run into cache invalidation problems.  And unless you do things carefully and laboriously in C, you'll waste a lot of space and leave performance on the table.  Supporting styles will be hard. External databases specialized for this purpose can use all the CPU and have optimized data structures.

**Reply by dgutov** [reply]
> And unless you do things carefully and laboriously in C, you'll waste a lot of space and leave performance on the table

Moving some hot path into C is usually an option, the main question is how to organize the logic on the high level.

> External databases specialized for this purpose can use all the CPU and have optimized data structures

How do *those* do cache invalidation?

**Reply by joaotavora** [reply]
> Moving some hot path into C is usually an option, the main question is how to organize the logic on the high level.

If that hot path includes anything related to Lisp completion styles, you've lost a big chunk of what makes having it in Emacs worth it.  And leveraging SMP and/or SIMD as I suspect these super-fast trendy rust alternatives do is still a big challenge.

> How do those do cache invalidation?

no idea, the point is it's not my problem.  as far as I can tell they work pretty well.

**Comment by joaotavora** [comment]
For what it's worth, I've recently bumped into a similar problem, and came up with this refinement of a previous solution in this discussion:

```lisp
(defvar eglot-lsp-context)
(defvar project-find-functions)
(require 'cl-lib)

(defun joaot/eglot-project-find-function (dir)
  (when eglot-lsp-context
    (let* ((marker nil)
           ;; We use "dominating files" here, but we could use some
           ;; other criteria, like an Elisp data structure, or hack in
           ;; dir-locals for DIR (maybe with
           ;; `hack-dir-local-variables-non-file-buffer').
           (root (locate-dominating-file
                  dir
                  (lambda (d)
                    ;; Here, we search for these markers regardless of
                    ;; other state such current major mode.  We could
                    ;; fix that easily, but that's out of scope for this small example.
                    (cl-loop for f in '("pyrightconfig.json" "pyproject.toml")
                             when (file-exists-p (expand-file-name f d))
			     do (setq marker f)
                             and return t)))))
      (when root
        ;; Use the experimental eglot--project project type in latest
        ;; Eglot which has some space for user-defined props,
        ;; `:marker' in this case.  One could make use of it, say, in a
        ;; functional entry to `eglot-server-programs', to decide
        ;; which server to run based on the marker or the type of
        ;; the project.
        `(eglot--project ,root :marker ,marker)))))

(add-to-list 'project-find-functions
             #'joaot/eglot-project-find-function)
```

It's just a small iteration on what was [already said above](https://github.com/joaotavora/eglot/discussions/1337#discussioncomment-8754812), just corrected to use `eglot-lsp-context`, use `dir` instead of the `default-directory` , and simplify the `throw/catch` (shouldn't really be needed if `locate-dominating-file` is doing it's job correctly).

**Reply by cowboyd** [reply]
Fwiw, I did a similar thing that used tsp-context and in addition to fixing the project context problems, it also solved the performance issues. I thought I had reported back on this, but apparently not. 

I'll see if I can adopt what you have here.
